File: experiments/java-sdk/src/main/java/com/datadog/llmobs/experiments/Dataset.java
package com.datadog.llmobs.experiments;

import com.datadog.api.client.v2.api.LlmObservabilityApi;
import com.datadog.api.client.v2.model.LLMObsDatasetDataAttributesRequest;
import com.datadog.api.client.v2.model.LLMObsDatasetDataRequest;
import com.datadog.api.client.v2.model.LLMObsDatasetRequest;
import com.datadog.api.client.v2.model.LLMObsDatasetResponse;
import com.datadog.api.client.v2.model.LLMObsDatasetType;
import com.datadog.llmobs.experiments.internal.DirectPost;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

/**
 * A dataset of records to run an experiment against.
 *
 * <p>Construct with {@link ExperimentsClient#createDataset(String)}. Add records with
 * {@link #addRecord}. The dataset is auto-created in Datadog and records are pushed when
 * {@link Experiment#run(int)} executes — no explicit push step in v0.1.
 */
public final class Dataset {

    private static final ObjectMapper JSON = new ObjectMapper();

    private final ExperimentsClient client;
    private final String name;
    private final String description;
    private final List<DatasetRecord> records = new ArrayList<>();
    private final List<String> recordIds = new ArrayList<>();

    private String id;
    private String projectId;
    private int pushedCount;

    Dataset(ExperimentsClient client, String name, String description) {
        this.client = Objects.requireNonNull(client, "client");
        this.name = Objects.requireNonNull(name, "name");
        this.description = description == null ? "" : description;
    }

    /**
     * Construct a Dataset that already exists in Datadog, pre-populated with records.
     * Package-private — used by {@link ExperimentsClient#pullDataset(String)}.
     */
    static Dataset fromExisting(
        ExperimentsClient client,
        String name,
        String description,
        String datasetId,
        String projectId,
        List<DatasetRecord> records,
        List<String> recordIds
    ) {
        Dataset d = new Dataset(client, name, description);
        d.id = datasetId;
        d.projectId = projectId;
        d.records.addAll(records);
        d.recordIds.addAll(recordIds);
        d.pushedCount = records.size();
        return d;
    }

    /**
     * Idempotently create this dataset in Datadog (if not already) and push any locally-buffered
     * records that haven't been pushed yet. Safe to call multiple times — only new records are
     * pushed on subsequent calls.
     */
    public void push() {
        ensureCreatedAndPushed(client.ensureProjectId());
    }

    public Dataset addRecord(DatasetRecord record) {
        Objects.requireNonNull(record, "record");
        records.add(record);
        return this;
    }

    public Dataset addRecord(Object input, Object expectedOutput) {
        return addRecord(new DatasetRecord(input, expectedOutput));
    }

    public Dataset addRecord(Object input, Object expectedOutput, Map<String, Object> metadata) {
        return addRecord(new DatasetRecord(input, expectedOutput, metadata));
    }

    public String name() {
        return name;
    }

    public List<DatasetRecord> records() {
        return Collections.unmodifiableList(records);
    }

    /** Datadog record IDs in the same order as {@link #records()}. Empty entries for any
     * records whose IDs we failed to parse from the push response. Package-private —
     * consumed by {@link Experiment#run()}. */
    List<String> recordIds() {
        return recordIds;
    }

    /** Datadog dataset ID. Null until the dataset has been created in Datadog. */
    public String id() {
        return id;
    }

    /** Datadog project ID. Null until the dataset has been pushed under a project. */
    public String projectId() {
        return projectId;
    }

    /**
     * Idempotently creates the dataset in Datadog under the given project and pushes any
     * pending local records. Called by {@link Experiment#run(int)}; users do not call this.
     */
    void ensureCreatedAndPushed(String projectId) {
        this.projectId = projectId;
        LlmObservabilityApi api = client.api();

        if (id == null) {
            LLMObsDatasetRequest req = new LLMObsDatasetRequest(
                new LLMObsDatasetDataRequest()
                    .type(LLMObsDatasetType.DATASETS)
                    .attributes(new LLMObsDatasetDataAttributesRequest(name).description(description))
            );
            try {
                LLMObsDatasetResponse resp = api.createLLMObsDataset(projectId, req);
                this.id = resp.getData().getId();
            } catch (Exception e) {
                throw new RuntimeException("Failed to create dataset '" + name + "': " + e.getMessage(), e);
            }
        }

        // Push only newly-added records (idempotent across multiple run() calls).
        if (pushedCount < records.size()) {
            // Hand-rolled HTTP: the generated client sends "type":"records" but the API
            // expects "type":"datasets" for this child-resource endpoint. Same spec/API
            // mismatch we hit for experiment events. Delete once aligned.
            List<DatasetRecord> toPush = records.subList(pushedCount, records.size());
            List<Map<String, Object>> items = new ArrayList<>(toPush.size());
            for (DatasetRecord r : toPush) {
                Map<String, Object> item = new LinkedHashMap<>();
                item.put("input", r.input());
                if (r.expectedOutput() != null) {
                    item.put("expected_output", r.expectedOutput());
                }
                if (!r.metadata().isEmpty()) {
                    item.put("metadata", r.metadata());
                }
                items.add(item);
            }
            Map<String, Object> attributes = new HashMap<>();
            attributes.put("records", items);

            Map<String, Object> body = new HashMap<>();
            Map<String, Object> data = new HashMap<>();
            data.put("type", "datasets");
            data.put("attributes", attributes);
            body.put("data", data);

            String path = "/api/v2/llm-obs/v1/" + projectId + "/datasets/" + id + "/records";
            String responseBody;
            try {
                responseBody = DirectPost.postReturning(
                    client.site(), path, client.apiKey(), client.applicationKey(), body
                );
            } catch (Exception e) {
                throw new RuntimeException("Failed to push records to dataset '" + name + "': " + e.getMessage(), e);
            }
            try {
                JsonNode dataNode = JSON.readTree(responseBody).path("data");
                if (dataNode.isArray()) {
                    for (JsonNode rec : dataNode) {
                        recordIds.add(rec.path("id").asText(""));
                    }
                }
            } catch (Exception e) {
                // Non-fatal: we lose dataset_record_id tagging for these rows, but the
                // experiment still runs. Log + continue with empty IDs for the new rows.
                int gap = (records.size() - pushedCount) - (recordIds.size() - pushedCount);
                for (int i = 0; i < gap; i++) {
                    recordIds.add("");
                }
            }
            pushedCount = records.size();
        }
    }
}

File: experiments/node-sdk/src/Dataset.ts
import { DatasetRecord } from "./DatasetRecord.js";
import type { ExperimentsClient } from "./ExperimentsClient.js";
import * as http from "./internal/http.js";
import { appBase } from "./internal/http.js";

/**
 * A local buffer of dataset records, auto-created and pushed remotely on the
 * first `experiment.run()` (or eagerly via `push()`). Mirrors the Java `Dataset`.
 *
 * `addRecord` is fluent. `id()`/`projectId()` are null until the dataset has been
 * pushed.
 */
export class Dataset {
  private readonly client: ExperimentsClient;
  private readonly _name: string;
  private readonly _description: string;
  private readonly _records: DatasetRecord[] = [];
  private readonly _recordIds: string[] = [];
  private _id: string | null = null;
  private _projectId: string | null = null;
  private pushedCount = 0;

  constructor(client: ExperimentsClient, name: string, description = "") {
    this.client = client;
    this._name = name;
    this._description = description;
  }

  /** @internal — build a Dataset that already exists remotely (used by pullDataset). */
  static fromExisting(
    client: ExperimentsClient,
    name: string,
    description: string,
    id: string,
    projectId: string,
    records: DatasetRecord[],
    recordIds: string[],
  ): Dataset {
    const ds = new Dataset(client, name, description);
    ds._id = id;
    ds._projectId = projectId;
    ds._records.push(...records);
    ds._recordIds.push(...recordIds);
    ds.pushedCount = records.length;
    return ds;
  }

  /** Append a record. Accepts either a DatasetRecord or `(input, expectedOutput?, metadata?)`. */
  addRecord(record: DatasetRecord): Dataset;
  addRecord(
    input: unknown,
    expectedOutput?: unknown,
    metadata?: Record<string, unknown>,
  ): Dataset;
  addRecord(
    recordOrInput: DatasetRecord | unknown,
    expectedOutput?: unknown,
    metadata?: Record<string, unknown>,
  ): Dataset {
    const record =
      recordOrInput instanceof DatasetRecord
        ? recordOrInput
        : new DatasetRecord(recordOrInput, expectedOutput, metadata);
    this._records.push(record);
    return this;
  }

  name(): string {
    return this._name;
  }

  records(): readonly DatasetRecord[] {
    return [...this._records];
  }

  /** @internal — record ids aligned by index with `records()` (after push). */
  recordIds(): readonly string[] {
    return [...this._recordIds];
  }

  /** Dataset id, or null until pushed. */
  id(): string | null {
    return this._id;
  }

  /** Project id, or null until pushed. */
  projectId(): string | null {
    return this._projectId;
  }

  /** Dashboard URL for this dataset, or null until it has been pushed/pulled. */
  url(): string | null {
    if (this._id === null) return null;
    return `${appBase(this.client.site())}/llm/datasets/${this._id}`;
  }

  /** Eagerly create the dataset (if needed) and push any unpushed records. */
  async push(): Promise<void> {
    const projectId = await this.client.ensureProjectId();
    await this.ensureCreatedAndPushed(projectId);
  }

  /**
   * @internal — create the remote dataset if it does not yet exist, then push any
   * records added since the last push. Idempotent and incremental: tracks
   * `pushedCount` so repeated calls only send new records.
   */
  async ensureCreatedAndPushed(projectId: string): Promise<void> {
    const creds = this.client.credentials();

    if (this._id === null) {
      // Dataset creation goes through the generated client (no active workaround).
      try {
        const resp = await this.client.api().createLLMObsDataset({
          projectId,
          body: {
            data: {
              type: "datasets",
              attributes: { name: this._name, description: this._description },
            },
          },
        });
        this._id = String(resp?.data?.id ?? "");
        this._projectId = projectId;
      } catch (err) {
        throw new Error(`Failed to create dataset '${this._name}': ${(err as Error).message}`);
      }
    }

    if (this.pushedCount >= this._records.length) {
      return;
    }

    const pending = this._records.slice(this.pushedCount);
    const recordPayload = pending.map((rec) => {
      const r: Record<string, unknown> = { input: rec.input };
      if (rec.expectedOutput !== null && rec.expectedOutput !== undefined) {
        r.expected_output = rec.expectedOutput;
      }
      if (rec.metadata && Object.keys(rec.metadata).length > 0) {
        r.metadata = rec.metadata;
      }
      return r;
    });

    // W1: the records POST must send type "datasets"; the generated client's
    // model still serializes "records", which the API rejects. Hand-roll it.
    const body = {
      data: {
        type: "datasets",
        attributes: { records: recordPayload },
      },
    };

    let raw: string;
    try {
      raw = await http.postReturning(
        creds,
        `/api/v2/llm-obs/v1/${projectId}/datasets/${this._id}/records`,
        body,
      );
    } catch (err) {
      throw new Error(`Failed to push records to dataset '${this._name}': ${(err as Error).message}`);
    }

    // Capture returned record ids in order; pad with "" if the response omits them.
    const parsed = JSON.parse(raw);
    const data = parsed?.data;
    if (Array.isArray(data)) {
      for (const node of data) {
        this._recordIds.push(String((node as Record<string, any>)?.id ?? ""));
      }
    } else {
      for (let i = 0; i < pending.length; i++) {
        this._recordIds.push("");
      }
    }

    this.pushedCount = this._records.length;
  }
}
